winpr/libwinpr/sspi/Negotiate/negotiate.c

```diff
@@ -679,36 +563,50 @@ static SECURITY_STATUS SEC_ENTRY negotiate_InitializeSecurityContextW(
 
 	if (!context)
 	{
+		enc = WinPrAsn1Encoder_New(WINPR_ASN1_DER);
+		if (!enc)
+			return SEC_E_INTERNAL_ERROR;
+
+		if (!WinPrAsn1EncSeqContainer(enc))
+			goto cleanup;
+
 		for (size_t i = 0; i < MECH_COUNT; i++)
 		{
 			MechCred* cred = &creds[i];
 
 			if (!cred->valid)
 				continue;
 
-			inner_mech_list_len += asn_tlv_length(cred->mech->oid->length);
-
-			if (init_context.mech) /* We already have an optimistic mechanism */
-				continue;
+			/* Send an optimistic token for the first valid mechanism */
+			if (!init_context.mech)
+			{
+				/* Use the output buffer to store the optimistic token */
+				CopyMemory(&output_token.mechToken, output_buffer, sizeof(SecBuffer));
 
-			/* Use the output buffer to store the optimistic token */
-			CopyMemory(&output_token.mechToken, output_buffer, sizeof(SecBuffer));
+				status = MechTable[i].pkg->table_w->InitializeSecurityContextW(
+				    &cred->cred, NULL, pszTargetName, fContextReq | cred->mech->flags, Reserved1,
+				    TargetDataRep, NULL, Reserved2, &init_context.sub_context, &mech_output,
+				    pfContextAttr, ptsExpiry);
 
-			status = MechTable[i].pkg->table_w->InitializeSecurityContextW(
-			    &cred->cred, NULL, pszTargetName, fContextReq | cred->mech->flags, Reserved1,
-			    TargetDataRep, NULL, Reserved2, &init_context.sub_context, &mech_output,
-			    pfContextAttr, ptsExpiry);
+				/* If the mechanism failed we can't use it; skip */
+				if (IsSecurityStatusError(status))
+				{
+					cred->valid = FALSE;
+					continue;
+				}
 
-			/* If the mechanism failed we can't use it; skip */
-			if (IsSecurityStatusError(status))
-				cred->valid = FALSE;
-			else
 				init_context.mech = cred->mech;
+				status = SEC_E_INTERNAL_ERROR;
```

Comment:
We managed to retrieve a security context, and the status is set to an error, not sure to understand...